File: frontend/src/games/mental-rotation/core/rng.ts
/* psygames-mental-rotation-rng · VER 1 · 23.08.2026 */
/**
 * СЛУЧАЙНОСТЬ, КОТОРУЮ МОЖНО ПОВТОРИТЬ.
 *
 * В бою пробы строятся на `Math.random`, и это правильно: одинаковые задания
 * подряд человек запоминает. Но проверить геометрию на «случайно повезло» нельзя:
 * упавшая раз в сотню проб подделка, совпавшая с правильным ответом, на прогоне
 * не воспроизведётся. Поэтому каждый строитель задания принимает `Rng`, а тесты
 * гоняют сотни проб по семени и падают на конкретном, названном в отчёте.
 *
 * Генератор — mulberry32 (тот же, что в «Соедини точки»): свой велосипед здесь
 * не нужен, нужен одинаковый ответ на одинаковое семя.
 */
import type { Rng } from './types';

export function hashSeed(seed: string): number {
  let hash = 0x811c9dc5;
  for (let index = 0; index < seed.length; index += 1) {
    hash ^= seed.charCodeAt(index);
    hash = Math.imul(hash, 0x01000193);
  }
  return hash >>> 0;
}

export function createRng(seed: string): Rng {
  let state = hashSeed(seed) || 1;
  return () => {
    state |= 0;
    state = (state + 0x6d2b79f5) | 0;
    let value = Math.imul(state ^ (state >>> 15), 1 | state);
    value = (value + Math.imul(value ^ (value >>> 7), 61 | value)) ^ value;
    return ((value ^ (value >>> 14)) >>> 0) / 4294967296;
  };
}

export function randomInt(rng: Rng, min: number, max: number): number {
  if (max < min) throw new RangeError(`пустой диапазон: ${min}..${max}`);
  return min + Math.floor(rng() * (max - min + 1));
}

export function pick<T>(rng: Rng, values: readonly T[]): T {
  if (values.length === 0) throw new RangeError('выбор из пустого набора');
  return values[randomInt(rng, 0, values.length - 1)] as T;
}

export function shuffle<T>(rng: Rng, values: readonly T[]): T[] {
  const result = [...values];
  for (let index = result.length - 1; index > 0; index -= 1) {
    const target = randomInt(rng, 0, index);
    [result[index], result[target]] = [result[target] as T, result[index] as T];
  }
  return result;
}

File: flutter/lib/games/mental_rotation/rng.dart
/// СЛУЧАЙНОСТЬ «МЫСЛЕННОГО ВРАЩЕНИЯ» — ТОТ ЖЕ ПОТОК, ЧТО В TS.
///
/// Перенос `frontend/src/games/mental-rotation/core/rng.ts` (mulberry32 + FNV-1a).
/// Совпадение обязано быть ПОБИТОВЫМ: задания сравниваются с эталонами, выгруженными прогоном
/// живого TS, а любое расхождение в младшем бите даёт другую фигуру и другой порядок вариантов.
///
/// 🔴 ДЖАВАСКРИПТ СЧИТАЕТ В 32 БИТАХ, DART — В 64. Поэтому каждое умножение здесь сначала режется
/// до 32 бит со знаком (`_imul`), а сдвиги вправо делаются по БЕЗЗНАКОВОМУ значению — иначе
/// отрицательное состояние размножит единицы слева, и потоки разойдутся на втором-третьем броске.
library;

/// Случайное число [0,1) — тот же вид, что у `Rng` в TS.
typedef Rng = double Function();

int _imul(int a, int b) => ((a.toSigned(32) * b.toSigned(32)) & 0xFFFFFFFF).toSigned(32);

/// FNV-1a по кодовым единицам UTF-16 — как `charCodeAt` в TS.
int hashSeed(String seed) {
  var hash = 0x811c9dc5.toSigned(32);
  for (final unit in seed.codeUnits) {
    hash = (hash ^ unit).toSigned(32);
    hash = _imul(hash, 0x01000193);
  }
  return hash & 0xFFFFFFFF;
}

Rng createRng(String seed) {
  var state = hashSeed(seed);
  if (state == 0) state = 1;
  state = state.toSigned(32);
  return () {
    state = (state + 0x6d2b79f5).toSigned(32);
    var value = _imul(state ^ ((state & 0xFFFFFFFF) >> 15), 1 | state);
    value = (value + _imul(value ^ ((value & 0xFFFFFFFF) >> 7), 61 | value)).toSigned(32) ^ value;
    return ((value ^ ((value & 0xFFFFFFFF) >> 14)) & 0xFFFFFFFF) / 4294967296;
  };
}

int randomInt(Rng rng, int min, int max) {
  if (max < min) throw RangeError('пустой диапазон: $min..$max');
  return min + (rng() * (max - min + 1)).floor();
}

T pick<T>(Rng rng, List<T> values) {
  if (values.isEmpty) throw RangeError('выбор из пустого набора');
  return values[randomInt(rng, 0, values.length - 1)];
}

List<T> shuffle<T>(Rng rng, List<T> values) {
  final result = [...values];
  for (var index = result.length - 1; index > 0; index -= 1) {
    final target = randomInt(rng, 0, index);
    final swap = result[index];
    result[index] = result[target];
    result[target] = swap;
  }
  return result;
}
